### PR TITLE
Ports/guile: Fix guile port build

### PR DESCRIPTION
**Pull request by petelliott**
image

doesn't change that you need the same version of guile to cross compile (pointed out in #19848). but it will build now, and I updated it to the latest version.

### CODE REVIEW

**Comment by BenWiederhake** [comment]
Wow, that was quick!

**Review by linusg — APPROVED** [review]

File: Userland/Libraries/LibC/complex.h [hunk 1]
```diff
@@ -74,4 +75,64 @@ static inline long double(cimagl)(long double complex z)
     return cimagl(z);
 }
 
+// https://pubs.opengroup.org/onlinepubs/9699919799/functions/cabs.html
+static inline long double cabsl(long double complex z)
+{
+    // Algorithm 312: Absolute value and square root of a complex number
+    // https://dl.acm.org/doi/10.1145/363717.363780
```
Comment (BenWiederhake) [inline Userland/Libraries/LibC/complex.h:82]: Thanks for adding that link! :D

File: Userland/Libraries/LibC/complex.h [hunk 2]
```diff
@@ -74,4 +75,64 @@ static inline long double(cimagl)(long double complex z)
...
+    if (real == 0 && imag == 0)
+        return CMPLXL(0, 0);
+
+    long double a = sqrtl((fabsl(real) + cabsl(z)) * 0.5);
+
+    if (real >= 0)
+        return CMPLXL(a, imag / (a + a));
+
+    long double b = (imag < 0) ? -a : a;
+    return CMPLXL(imag / (b + b), b);
```
Comment (BenWiederhake) [inline Userland/Libraries/LibC/complex.h:125]: This looks like a faithful implementation of the paper, yes.

However, real and imag can still "round to zero" while computing `a`, for example when called with ridiculously small values like the subnormals. How about we don't check for `real == 0.0 && imag == 0.0`. Instead, I suggest:

```c++
    long double a_squared = (fabsl(real) + cabsl(z)) * 0.5;
    if (a_squared <= 0.0) { return CMPLXL(0, 0); }
    // Square-rooting a positive number makes it get closer to 1, so `a` is clearly non-zero.
    long double a = sqrtl(a_squared);
    // Continue as before
```

Can you also add tests to make sure that there were no little typos or other mistakes? Something like:
* 1
* -1
* 2
* 3i
* -9i
* -7+24i which should result in "exactly" 3+4i
* 7+24i which should result in "exactly" 4+3i
* -9+0.00001i which should be somewhat close to 0+3i
* -9-0.00001i which should be somewhat close to 0-3i. Note that if the implementation splits the sqrt-branch in a different place, this should be documented
* 1e-4950 `== std::numeric_limits<long double>::denorm_min()` in STL, I forgot how we do it in AK. The result should be something like `6.03755e-2476`.

I got the last two numbers using
```c++
// g++ foo.cpp -Wall -pedantic -o foo && ./foo
#include <cmath>
#include <iostream>
#include <limits>
int main() {
    std::cout << std::numeric_limits<long double>::denorm_min() << '\n';
    std::cout << sqrtl(std::numeric_limits<long double>::denorm_min()) << '\n';
}
```

**Review by BenWiederhake — CHANGES_REQUESTED** [review]
Looks good, and thank you so much for the spec links :D

I have some concerns about the new math functions, please implement some tests for them. I believe there's a typo.

File: Userland/Libraries/LibC/complex.h [hunk 3]
```diff
@@ -74,4 +75,64 @@ static inline long double(cimagl)(long double complex z)
...
+    // Algorithm 312: Absolute value and square root of a complex number
+    // https://dl.acm.org/doi/10.1145/363717.363780
+    long double real = creal(z);
+    long double imag = cimagl(z);
+
+    if (real == 0)
+        return imag;
+
+    if (imag == 0)
+        return real;
```
[the remark this answers is not in the document]
Comment (petelliott) [reply Userland/Libraries/LibC/complex.h:90]: yep, this should operate on the absolute values of the real and imaginary parts

**Comment by BenWiederhake** [comment]
Sorry linus, but I hope the situation is clear enough that I can change the label, after all the pr author also agrees.

File: Userland/Libraries/LibC/complex.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (petelliott) [reply Userland/Libraries/LibC/complex.h:92]: fixed

File: Userland/Libraries/LibC/complex.h (hunk 2 above)
Comment (petelliott) [reply Userland/Libraries/LibC/complex.h:129]: changed the logic, and added the tests

File: Tests/LibC/TestCComplex.c [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+                __FILE__, __LINE__, #a, #b);                                            \
+            fail_counter++;                                                             \
+        }                                                                               \
+    }
+
+#define EXPECT_APPROXIMATE_WITH_ERROR(a, b, err)                                                 \
+    {                                                                                            \
+        typeof(a) lhs = a;                                                                       \
+        typeof(b) rhs = b;                                                                       \
+        if (cabsl(lhs - rhs) > err) {                                                            \
```
Comment (BenWiederhake) [inline Tests/LibC/TestCComplex.c:34]: This assumes that `cabsl` already works while testing `cabsl`. A silly implementation like `double cabsl(complex foo) { return -1.0; }` would probably pass this test.

Can this be avoided? What do you think about something like this:
`if (absl(creall(a) - creall(b)) > err || absl(cimagl(a) - cimagl(b)) > err) {`

File: Userland/Libraries/LibC/complex.h [hunk 5]
```diff
@@ -11,10 +11,13 @@
 
 #pragma once
 
+/*
```
Comment (BenWiederhake) [inline Userland/Libraries/LibC/complex.h:14]: Why is this commented out? :thinking:

**Review by BenWiederhake — CHANGES_REQUESTED** [review]
Thank you! :D

Because this now uses a cmake feature that looks really exotic in my eyes, we should probably add a comment why we need it. Also, one of the tests relies on the thing under test already working, let's not do that :D

File: Tests/LibC/TestCComplex.c (hunk 4 above)
Comment (petelliott) [reply Tests/LibC/TestCComplex.c]: done

File: Meta/CMake/common_compile_options.cmake [hunk 6]
```diff
@@ -27,8 +27,8 @@ if (CMAKE_CXX_COMPILER_ID MATCHES "Clang$")
 elseif (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
     # Only ignore expansion-to-defined for g++, clang's implementation doesn't complain about function-like macros
     add_compile_options(-Wno-expansion-to-defined)
-    add_compile_options(-Wno-literal-suffix)
+    add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-Wno-literal-suffix>)
```
[the remark this answers is not in the document]
Comment (petelliott) [reply Meta/CMake/common_compile_options.cmake]: comments added in both files that i added these

**Review by BenWiederhake — APPROVED** [review]
Looks good, let's merge it :D

File: Tests/LibC/TestCComplex.c
Comment (linusg) [inline Tests/LibC/TestCComplex.c:1]: This doesn't build on CI

**Review by linusg — CHANGES_REQUESTED** [review]

**Pull request closed without merging**